### PR TITLE
feat: add producer

### PR DESCRIPTION
**Pull request by cyberhck**
## Pull Request Submission Checklist

Please confirm that you have done the following before requesting reviews:

- [x] I have confirmed that the PR type is appropriate for the change I am making according to
  the [Honest Pull Request and Commit Message Naming Conventions](https://www.notion.so/honestbank/Pull-Request-and-Commit-Message-Naming-Conventions-bd97f2cbb34c4c73b1ff3a3e384b850c).
- [x] I have typed an adequate description that explains **why** I am making this change.
- [x] I have installed and run standard pre-commit hooks that lints and validates my code.

### Description

* adding producer API

- - -
This change is [Reviewable](https://reviewable.io/reviews/honestbank/kp/16)

### CODE REVIEW

**Review by kelu-diao — COMMENTED** [review]
Nicely done! I think this is a decent implementation :)

File: v2/internal/schemaregistry/config.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package schemaregistry
+
+type Config struct {
+	Endpoint string `env:"KP_SCHEMA_REGISTRY_ENDPOINT" required:"true"`
+	Username string `env:"KP_SCHEMA_REGISTRY_USERNAME" default:""`
```
Comment (bludot) [inline v2/internal/schemaregistry/config.go:5]: wouldnt this also be required? can help reduce mistakes?

File: v2/internal/serialization/decoder.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+	"github.com/heetch/avro"
+)
+
+func Decode[MessageType any](payload []byte) (*MessageType, error) {
+	var emptyMessage MessageType
+	typ, err := avro.TypeOf(emptyMessage)
+	if err != nil {
+		return nil, err
+	}
+	_, err = avro.Unmarshal(payload[5:], &emptyMessage, typ)
```
Comment (bludot) [inline v2/internal/serialization/decoder.go:13]: what is this static `5` for? maybe add a comment?

File: v2/producer/producer.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	msg := &kafka.Message{
+		TopicPartition: partition,
+		Value:          value,
+	}
+
+	return p.k.Produce(msg, nil)
+}
+
+func (p producer[BodyType, KeyType]) Flush() error {
+	p.k.Flush(3000)
```
Comment (bludot) [inline v2/producer/producer.go:37]: what is the static `3000` for?

**Review by bludot — CHANGES_REQUESTED** [review]

File: v2/internal/schemaregistry/config.go (hunk 1 above)
Comment (cyberhck) [reply v2/internal/schemaregistry/config.go:5]: in some cases you don't need username and passwords (like while working locally)

**Review by bludot — APPROVED** [review]

File: v2/producer/producer.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+	cfg = hydrateIfNotNil(cfg, "bootstrap.servers", kafkaConfig.BootstrapServers)
+	cfg = hydrateIfNotNil(cfg, "sasl.mechanisms", kafkaConfig.SaslMechanism)
+	cfg = hydrateIfNotNil(cfg, "security.protocol", kafkaConfig.SecurityProtocol)
+	cfg = hydrateIfNotNil(cfg, "sasl.username", kafkaConfig.Username)
+
+	return hydrateIfNotNil(cfg, "sasl.password", kafkaConfig.Password)
+}
+
+func hydrateIfNotNil(cfg *kafka.ConfigMap, key string, value *string) *kafka.ConfigMap {
```
Comment (poom) [inline v2/producer/producer.go:74]: It is pointer. I think it doesn't need to return

File: v2/producer/producer.go [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+func getKafkaConfig(kafkaConfig config.KafkaConfig) *kafka.ConfigMap {
+	cfg := &kafka.ConfigMap{}
+
+	cfg = hydrateIfNotNil(cfg, "bootstrap.servers", kafkaConfig.BootstrapServers)
+	cfg = hydrateIfNotNil(cfg, "sasl.mechanisms", kafkaConfig.SaslMechanism)
+	cfg = hydrateIfNotNil(cfg, "security.protocol", kafkaConfig.SecurityProtocol)
+	cfg = hydrateIfNotNil(cfg, "sasl.username", kafkaConfig.Username)
+
+	return hydrateIfNotNil(cfg, "sasl.password", kafkaConfig.Password)
```
Comment (poom) [inline v2/producer/producer.go:71]:
```suggestion
	hydrateIfNotNil(cfg, "bootstrap.servers", kafkaConfig.BootstrapServers)
	hydrateIfNotNil(cfg, "sasl.mechanisms", kafkaConfig.SaslMechanism)
	hydrateIfNotNil(cfg, "security.protocol", kafkaConfig.SecurityProtocol)
	hydrateIfNotNil(cfg, "sasl.username", kafkaConfig.Username)
	hydrateIfNotNil(cfg, "sasl.password", kafkaConfig.Password)

	return cfg
```

File: v2/producer/producer.go (hunk 4 above)
Comment (poom) [reply v2/producer/producer.go:75]: @cyberhck change func signature too

**Review by poom — CHANGES_REQUESTED** [review]
check my comment

**Review by poom — APPROVED** [review]

**Review by jai — APPROVED** [review]
> Reviewed 11 of 18 files at r1, 1 of 3 files at r2, 1 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/honestbank/kp/16)* status: 13 of 18 files reviewed, 6 unresolved discussions (waiting on @bludot, @cyberhck, and @poom)

___
*[`v2/internal/serialization/decoder.go` line 13 at r3](https://reviewable.io/reviews/honestbank/kp/16#-NFWjloR81vqBFkqoxGX:-NFWjloR81vqBFkqoxGY:undefined) ([raw file](https://github.com/honestbank/kp/blob/e027597590838af9538aa4c755bce6858b2bca86/v2/internal/serialization/decoder.go#L13)):*
> ```Go
> 		return nil, err
> 	}
> 	// first byte is the magic byte and the 4 bytes are the schema ID int which is why we start reading from 5th position
> ```





_Suggestion:_
```Go
	// first byte is the magic byte and the next 4 bytes are the schema ID (int) which is why we start reading from the 5th position
```

File: v2/internal/serialization/decoder.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	"github.com/heetch/avro"
+)
+
+func Decode[MessageType any](payload []byte) (*MessageType, error) {
+	var emptyMessage MessageType
+	typ, err := avro.TypeOf(emptyMessage)
+	if err != nil {
+		return nil, err
+	}
+	// first byte is the magic byte and the 4 bytes are the schema ID int which is why we start reading from 5th position
```
Comment (cyberhck) [inline v2/internal/serialization/decoder.go:13]:
```suggestion
	// first byte is the magic byte and the next 4 bytes are the schema ID (int) which is why we start reading from the 5th position
```

**Pull request merged**